File: app/src/androidTest/java/de/westnordost/streetcomplete/data/meta/CountryInfosTest.java
package de.westnordost.streetcomplete.data.meta;

import android.content.res.AssetManager;

import org.junit.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.westnordost.streetcomplete.quests.opening_hours.model.Weekdays;

import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
import static org.junit.Assert.*;

public class CountryInfosTest
{
	private void checkFirstDayOfWorkweekIsValid(CountryInfo info)
	{
		assertNotNull(info.getFirstDayOfWorkweek());
		assertTrue(Weekdays.Companion.getWeekdayIndex(info.getFirstDayOfWorkweek()) > -1);
	}

	private void checkLengthUnitIsEitherMeterOrFootAndInch(CountryInfo info)
	{
		assertNotNull(info.getLengthUnits());
		assertTrue(info.getLengthUnits().contains("meter") || info.getLengthUnits().contains("foot and inch"));
	}

	private void checkSpeedUnitIsEitherMphOrKmh(CountryInfo info)
	{
		assertNotNull(info.getSpeedUnits());
		assertTrue(info.getSpeedUnits().contains("kilometers per hour") || info.getSpeedUnits().contains("miles per hour"));
	}


	private void checkWeightLimitUnitIsEitherTonOrShortTonOrPound(CountryInfo info)
	{
		assertNotNull(info.getWeightLimitUnits());
		assertTrue(info.getWeightLimitUnits().contains("ton") || info.getWeightLimitUnits().contains("short ton") || info.getWeightLimitUnits().contains("pound"));
	}
	private void checkAdditionalValidHousenumberRegexes(Map<String, CountryInfo> infos)
	{
		assertTrue("99 bis".matches(infos.get("FR").getAdditionalValidHousenumberRegex()));
		assertTrue("s/n".matches(infos.get("ES").getAdditionalValidHousenumberRegex()));
	}

	private void checkRegularShoppingDaysIsBetween0And7(CountryInfo info)
	{
		assertNotNull(info.getRegularShoppingDays());
		assertTrue(info.getRegularShoppingDays() <= 7);
		assertTrue(info.getRegularShoppingDays() >= 0);
	}

	private static final List<String> validWeekdays = Arrays.asList("Mo","Tu","We","Th","Fr","Sa","Su");
	private void checkStartOfWorkweekValid(CountryInfo info)
	{
		assertTrue(validWeekdays.contains(info.getFirstDayOfWorkweek()));
	}

	private void checkForEach(CountryInfo info)
	{
		checkFirstDayOfWorkweekIsValid(info);
		checkLengthUnitIsEitherMeterOrFootAndInch(info);
		checkSpeedUnitIsEitherMphOrKmh(info);
		checkWeightLimitUnitIsEitherTonOrShortTonOrPound(info);
		checkRegularShoppingDaysIsBetween0And7(info);
		checkStartOfWorkweekValid(info);
	}

	@Test public void all() throws IOException
	{
		Map<String, CountryInfo> infos = getAllCountryInfos();
		for(Map.Entry<String,CountryInfo> elem : infos.entrySet())
		{
			try
			{
				CountryInfo ci = elem.getValue();
				assertEquals(elem.getKey().split("-")[0], ci.countryCode);
				checkForEach(ci);
			}
			catch (Throwable e)
			{
				throw new RuntimeException("Error for "+ elem.getKey(), e);
			}
		}

		checkAdditionalValidHousenumberRegexes(infos);
	}

	private Map<String, CountryInfo> getAllCountryInfos() throws IOException
	{
		AssetManager am = getInstrumentation().getTargetContext().getAssets();
		String[] fileList = am.list("country_metadata");
		assertNotNull(fileList);
		CountryInfos cis = new CountryInfos(am, null);
		Map<String,CountryInfo> all = new HashMap<>();
		for (String filename : fileList)
		{
			String country = filename.substring(0, filename.lastIndexOf("."));
			try
			{
				CountryInfo info = cis.get(Collections.singletonList(country));
				all.put(country, info);
			}
			catch (Throwable e)
			{
				throw new RuntimeException("Error for " + filename, e);
			}
		}
		return all;
	}
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/data/meta/CountryInfosTest.kt
package de.westnordost.streetcomplete.data.meta

import androidx.test.platform.app.InstrumentationRegistry
import de.westnordost.streetcomplete.osm.LengthUnit
import de.westnordost.streetcomplete.osm.opening_hours.model.Weekdays.Companion.getWeekdayIndex
import de.westnordost.streetcomplete.quests.max_speed.SpeedMeasurementUnit
import de.westnordost.streetcomplete.quests.max_weight.WeightMeasurementUnit
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.IOException

class CountryInfosTest {
    private val validWeekdays = listOf("Mo", "Tu", "We", "Th", "Fr", "Sa", "Su")

    private fun checkFirstDayOfWorkweekIsValid(info: CountryInfo) {
        assertNotNull(info.firstDayOfWorkweek)
        assertTrue(getWeekdayIndex(info.firstDayOfWorkweek) > -1)
        assertTrue(getWeekdayIndex(info.firstDayOfWorkweek) < 7)
    }

    private fun checkLengthUnitIsEitherMeterOrFootAndInch(info: CountryInfo) {
        assertNotNull(info.lengthUnits)
        assertTrue(info.lengthUnits.contains(LengthUnit.METER)
            || info.lengthUnits.contains(LengthUnit.FOOT_AND_INCH)
        )
    }

    private fun checkSpeedUnitIsEitherMphOrKmh(info: CountryInfo) {
        assertNotNull(info.speedUnits)
        assertTrue(info.speedUnits.contains(SpeedMeasurementUnit.KILOMETERS_PER_HOUR)
            || info.speedUnits.contains(SpeedMeasurementUnit.MILES_PER_HOUR)
        )
    }

    private fun checkWeightLimitUnitIsEitherTonOrShortTonOrPound(info: CountryInfo) {
        assertNotNull(info.weightLimitUnits)
        assertTrue(info.weightLimitUnits.contains(WeightMeasurementUnit.TON)
            || info.weightLimitUnits.contains(WeightMeasurementUnit.SHORT_TON)
            || info.weightLimitUnits.contains(WeightMeasurementUnit.POUND)
        )
    }

    private fun checkAdditionalValidHousenumberRegexes(infos: Map<String, CountryInfo>) {
        assertTrue("99 bis".matches(infos["FR"]!!.additionalValidHousenumberRegex!!.toRegex()))
        assertTrue("s/n".matches(infos["ES"]!!.additionalValidHousenumberRegex!!.toRegex()))
    }

    private fun checkRegularShoppingDaysIsBetween0And7(info: CountryInfo) {
        assertNotNull(info.regularShoppingDays)
        assertTrue(info.regularShoppingDays <= 7)
        assertTrue(info.regularShoppingDays >= 0)
    }

    private fun checkStartOfWorkweekValid(info: CountryInfo) {
        assertTrue(validWeekdays.contains(info.firstDayOfWorkweek))
    }

    @Test
    @Throws(IOException::class)
    fun all() {
        val infos = getAllCountryInfos()
        for ((key, countryInfo) in infos) {
            try {
                assertEquals(key.split("-".toRegex()).toTypedArray()[0], countryInfo.countryCode)
                checkFirstDayOfWorkweekIsValid(countryInfo)
                checkLengthUnitIsEitherMeterOrFootAndInch(countryInfo)
                checkSpeedUnitIsEitherMphOrKmh(countryInfo)
                checkWeightLimitUnitIsEitherTonOrShortTonOrPound(countryInfo)
                checkRegularShoppingDaysIsBetween0And7(countryInfo)
                checkStartOfWorkweekValid(countryInfo)
            } catch (e: Throwable) {
                throw RuntimeException("Error for $key", e)
            }
        }
        checkAdditionalValidHousenumberRegexes(infos)
    }

    @Throws(IOException::class)
    private fun getAllCountryInfos(): Map<String, CountryInfo> {
        val assetManager = InstrumentationRegistry.getInstrumentation().targetContext.assets
        val countryInfos = CountryInfos(assetManager, null)
        val fileList = assetManager.list("country_metadata")!!
        assertNotNull(fileList)

        return fileList.associate { filename ->
            val country = filename.substring(0, filename.lastIndexOf("."))
            try {
                (country to countryInfos.get(listOf(country)))
            } catch (e: Throwable) {
                throw RuntimeException("Error for $filename", e)
            }
        }
    }
}
